### PR TITLE
Void Analysis

### PR DESCRIPTION
**Pull request by GabrielBram**
**IMPLEMENTATION OVERVIEW**
- Implements a system of numpy meshgrid for the storage of volume data on a cartesian grid.
- Storage and definition of these grids defined in a Unit_Cell class.
- Functionality used to plot the volume of space that can be occupied by a probe sphere of a given diameter (void space).
- Outputs basic information such as the total volume of the void and the location and radius of the largest sphere space.
- Example scripts added for an MFI framework.  

**FUTURE WORK**
- Change the junk value for points not utilised on the meshgrid.
- More sophisticated algorithm (Delauney triangulation for example) to determine volume that is large enough to accommodate probe spheres, but inaccessible due to bottlenecks.
- Add functionality to read .cube data for general analysis. 
- Meshgrid system is generalisable, so flexible to other volume data used for other purposes.

### CODE REVIEW

File: examples/meshgrid_voidanalysis_MFI.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from carmm.meshgrid.meshgrid_unitcell import UnitCell
+from carmm.meshgrid.meshgrid_functions import atom_mesh_build_mask
+from carmm.meshgrid.meshgrid_void import void_find
+from carmm.meshgrid.meshgrid_void import void_build_mask
+from carmm.meshgrid.meshgrid_void import void_analysis
+import matplotlib.pyplot as plt
+from ase import io
+
+# An example showing how to plot the total volume not occupied by molecular vdW volume.
```
Comment (logsdail) [inline examples/meshgrid_voidanalysis_MFI.py:11]: For some reason with the QA testing (and don't ask why), test cases have to be setup in a function and then that function itself called. See e.g. https://github.com/logsdail/carmm/blob/master/examples/analyse_angles.py#L9 for a working example. I've no problem with the example here, but can it be changed into the format highlighted?

File: carmm/meshgrid/meshgrid_void.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    void_radii = []
+    atom_radii = []
+
+    for atom_co in range(len(atom.positions)):
+        at_symbols = atom.get_chemical_symbols()
+        at_n = atomic_numbers[at_symbols[atom_co]]
+        atom_radius = vdw_radii[at_n]
+        atom_radii.append(atom_radius)
+
+    atom_radii = np.array([atom_radii]).flatten()
```
Comment (PavelStishenko) [inline carmm/meshgrid/meshgrid_void.py:35]: I think it is equivalent to this code: `atom_radii = vdw_radii[atom.numbers]`. Am I missing something?

File: carmm/meshgrid/meshgrid_void.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    for i in range(0, Ucell.nx, coarseness):
+        for j in range(0, Ucell.ny, coarseness):
+            for k in range(0, Ucell.nz, coarseness):
```
Comment (PavelStishenko) [inline carmm/meshgrid/meshgrid_void.py:39]: I would use `itertools.product` or `np.ndindex` here. But is the matter of taste.

File: carmm/meshgrid/meshgrid_void.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+                a_xx, a_yy, a_zz = Ucell.xx[i, j, k], Ucell.yy[i, j, k], Ucell.zz[i, j, k]
+
+                distances = np.abs(atom.positions - np.array([a_xx, a_yy, a_zz]))
+                if mic:
+                    distances = np.where(distances > 0.5 * Ucell.dim, distances - Ucell.dim, distances)
+
+                distances = np.sqrt((distances ** 2).sum(axis=-1)) - atom_radii
```
Comment (PavelStishenko) [inline carmm/meshgrid/meshgrid_void.py:47]: Looks nice ;-)

File: carmm/meshgrid/meshgrid_void.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    for i in range(0, Ucell.nx, coarseness):
+        for j in range(0, Ucell.ny, coarseness):
+            for k in range(0, Ucell.nz, coarseness):
+
+                a_xx, a_yy, a_zz = Ucell.xx[i, j, k], Ucell.yy[i, j, k], Ucell.zz[i, j, k]
+
+                distances = np.abs(atom.positions - np.array([a_xx, a_yy, a_zz]))
+                if mic:
+                    distances = np.where(distances > 0.5 * Ucell.dim, distances - Ucell.dim, distances)
```
Comment (PavelStishenko) [inline carmm/meshgrid/meshgrid_void.py:45]: May be it is job for `ase.geometry.find_mic` too.  Isn't it?

File: carmm/meshgrid/meshgrid_void.py (hunk 4 above)
Comment (PavelStishenko) [inline carmm/meshgrid/meshgrid_void.py:47]: `distances = np.linalg.norm(distances, axis=-1) - atom_radii`

File: carmm/meshgrid/meshgrid_functions.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
+def distance_meshgrid2point(a_xx, a_yy, a_zz, xx, yy, zz, dim, mic):
```
[the remark this answers is not in the document]
Comment (GabrielBram) [reply carmm/meshgrid/meshgrid_functions.py:1]: Good idea - doing this currently but it requires a little bit of re-writing - will have this done by today I reckon.

File: carmm/meshgrid/meshgrid_void.py (hunk 2 above)
Comment (GabrielBram) [reply carmm/meshgrid/meshgrid_void.py:35]: You aren't missing anything - this was a round the houses approach for little reason.

File: carmm/meshgrid/meshgrid_void.py (hunk 3 above)
Comment (GabrielBram) [reply carmm/meshgrid/meshgrid_void.py:39]: Will bear this in mind for future, but for the sake of time will leave it as is in this code.

File: carmm/meshgrid/meshgrid_void.py (hunk 5 above)
Comment (GabrielBram) [reply carmm/meshgrid/meshgrid_void.py:45]: It is indeed! Swapping out this function for the ase.geometry equivalent.

File: carmm/meshgrid/meshgrid_unitcell.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        self.dim = 0
+
+    def define_unit_cell(self, x, y, z, nx=50, ny=50, nz=50):
+        """
+        Defines a unit cell of (nx, ny, nz) points of (x, y, z) length. Also defines the partition box indices
+        used for splitting up the simulation cell into smaller units.
+        Args:
+            n_x, n_y, n_z: int
+                Number of points along each cartesian axis.
+            x, y, z: float
```
[the remark this answers is not in the document]
Comment (GabrielBram) [reply carmm/meshgrid/meshgrid_unitcell.py:23]: I've just looked through again and the ase.geometry and get_distances functions in ase are very good for point to point distances, but making them behave with the meshgrids (unless I'm missing something) requires reshaping the meshgrids into a list of (Nx3) vectors. The result is significantly slower than the current implementation.

File: carmm/analyse/meshgrid_void.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        void_max: float
+            Maximum radius scanned from for probe sphere.
+        ucell_obj: unit_cell object
+            Contains meshgrid and associated unit cell information.
+        mol_xx: numpy array [nx,ny,nz]
+            Grid to determine whether molecule is present at a point.
+        mic: logical
+            Use miniumum image convention.
+        coarseness: integer
+            Skips over an integer number of points
```
Comment (ikowalec) [inline carmm/analyse/meshgrid_void.py]: atomco, atom_rad description missing

File: carmm/analyse/meshgrid/meshgrid_functions.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            y coordinate of point 1.
+        z_1: float
+            z coordinate of point 1.
+        x_2: float
+            x coordinate of point 2.
+        y_2: float
+            y coordinate of point 2.
+        z_2: float2
+            z coordinate of point 2.
+        MeshObject: MeshgridObject
```
Comment (ikowalec) [inline carmm/analyse/meshgrid/meshgrid_functions.py:86]: Is the change in naming scheme necessary?

File: carmm/analyse/meshgrid/meshgrid_functions.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+    midpoint = np.linalg.norm(vec1+(mic_shift/2))
+
+    return midpoint
+
+def atom_mesh_build_mask(MeshObject, Atom):
+    """
+    Defines meshgrid of based on a radius set by atomic_radius. Uses 99.99 as a junk value.
+    FUTURE DEVELOPMENT:
+    - Replace junk value with None??
```
Comment (ikowalec) [inline carmm/analyse/meshgrid/meshgrid_functions.py:113]: I second that

File: carmm/analyse/meshgrid/meshgrid_functions.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    Args:
+        MeshObject: MeshgridObject
+            Object storing meshgrid and PBC conditions
+        Atom: Atoms object
+            Contains information (atomic symbols and positions)
```
Comment (ikowalec) [inline carmm/analyse/meshgrid/meshgrid_functions.py:118]: There is an Atom class in ASE, the name can be confusing, change to atoms or Atoms

File: examples/analyse_meshgrid_voidanalysis_MFI.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+def test_void_find():
+
+    from carmm.analyse.meshgrid.meshgrid_meshobject import Mesh
+    from carmm.analyse.meshgrid.meshgrid_functions import atom_mesh_build_mask, distance_meshgrid2point
+    from carmm.analyse.meshgrid.meshgrid_void import void_find, void_build_mask, void_analysis
+    import matplotlib.pyplot as plt
+    import numpy as np
+    from ase.io import read
+
+    # Initialise the Mesh object, with a default underlying meshgrid of 50x50x50 grid points.
```
Comment (ikowalec) [inline examples/analyse_meshgrid_voidanalysis_MFI.py:22]: where does 50x50x50 come from? 20x20x20 below

File: examples/analyse_meshgrid_voidanalysis_MFI.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    mfi_mesh = Mesh(np.array([20.22614449, 19.82125040, 13.36948553, 90, 90, 90]),
+                    nx = 20, ny = 20, nz = 20, pbc=[1, 1, 1])
+
+    # Read atomic coordinates and create a meshgrid using periodic boundary conditions
+    # Meshgrid arrays print the x, y, z coordinates for all given points into three (nx,ny,nz)
+    # arrays (xx, yy and zz respectively).
+    atom = read('data/Zeolite/MFI_framework_geom.xyz')
```
Comment (ikowalec) [inline examples/analyse_meshgrid_voidanalysis_MFI.py:29]: Misleading variable name `atom` containing an `Atoms` object

File: carmm/analyse/meshgrid/meshgrid_void.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
+def void_find(MeshObject, Atom, coarseness=1):
+    """
+    Defines the void centres and radii of points across the meshgrid. Coarseness factor reduces number of probe points
+    by skipping over points in the grid. Uses 99.99 as a junk value.
+    Args:
+        MeshObject: MeshgridObject object
+            Contains meshgrid and associated unit cell information.
+        Atom: Atoms object
```
Comment (ikowalec) [inline carmm/analyse/meshgrid/meshgrid_void.py:8]: Atoms object is a list of Atom objects, misleading name

File: examples/analyse_meshgrid_voidanalysis_MFI.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    from carmm.analyse.meshgrid.meshgrid_meshobject import Mesh
+    from carmm.analyse.meshgrid.meshgrid_functions import atom_mesh_build_mask, distance_meshgrid2point
+    from carmm.analyse.meshgrid.meshgrid_void import void_find, void_build_mask, void_analysis
+    import matplotlib.pyplot as plt
+    import numpy as np
+    from ase.io import read
+
+    # Initialise the Mesh object, with a default underlying meshgrid of 50x50x50 grid points.
+    mfi_mesh = Mesh(np.array([20.22614449, 19.82125040, 13.36948553, 90, 90, 90]),
+                    nx = 20, ny = 20, nz = 20, pbc=[1, 1, 1])
```
Comment (ikowalec) [inline examples/analyse_meshgrid_voidanalysis_MFI.py:24]: What if there is a mismatch between Atoms.pbc and Mesh.pbc?

File: examples/analyse_meshgrid_voidanalysis_MFI.py (hunk 12 above)
Comment (GabrielBram) [reply examples/analyse_meshgrid_voidanalysis_MFI.py:22]: Ah, I changed this to make the test run faster, but the comment does not reflect this. Changed as appropriate.

File: carmm/analyse/meshgrid/meshgrid_functions.py (hunk 11 above)
Comment (GabrielBram) [reply carmm/analyse/meshgrid/meshgrid_functions.py:118]: Quite right. Resolved.

File: examples/analyse_meshgrid_voidanalysis_MFI.py (hunk 13 above)
Comment (GabrielBram) [reply examples/analyse_meshgrid_voidanalysis_MFI.py:29]: Good point. Resolved.

File: carmm/analyse/meshgrid/meshgrid_void.py (hunk 14 above)
Comment (GabrielBram) [reply carmm/analyse/meshgrid/meshgrid_void.py:8]: Ditto. Resolved.

File: examples/analyse_meshgrid_voidanalysis_MFI.py (hunk 15 above)
Comment (GabrielBram) [reply examples/analyse_meshgrid_voidanalysis_MFI.py]: True - have included a function that checks whether PBCs match (and an option to swtich it off for future bug fixing) wherever atoms objects are used.

File: carmm/analyse/meshgrid/meshgrid_functions.py (hunk 10 above)
Comment (GabrielBram) [reply carmm/analyse/meshgrid/meshgrid_functions.py]: Used np.nan, but quite the same effect achieved - more robust than setting an arbitrary value.

File: carmm/analyse/meshgrid/meshgrid_functions.py (hunk 9 above)
Comment (GabrielBram) [reply carmm/analyse/meshgrid/meshgrid_functions.py]: Not at all - resolved.

**Pull request merged**